From 479c5ab17bdd9ae592f678f76638334778b7c3e1 Mon Sep 17 00:00:00 2001
From: Brad Geltz <brad.geltz@intel.com>
Date: Mon, 24 Aug 2020 17:14:19 -0700
Subject: [PATCH 8/8] Add barriers to collective calls
#
#  Copyright (c) 2015, 2016, 2017, 2018, 2019, 2020, Intel Corporation
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions
#  are met:
#
#      * Redistributions of source code must retain the above copyright
#        notice, this list of conditions and the following disclaimer.
#
#      * Redistributions in binary form must reproduce the above copyright
#        notice, this list of conditions and the following disclaimer in
#        the documentation and/or other materials provided with the
#        distribution.
#
#      * Neither the name of Intel Corporation nor the names of its
#        contributors may be used to endorse or promote products derived
#        from this software without specific prior written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#

- Add special makefile to set build flag which enables
  barrier usage before collectives.

Signed-off-by: Brad Geltz <brad.geltz@intel.com>
---
 src/comm_mpi.f              | 9 +++++++++
 test/example1/makenek-intel | 3 +++
 2 files changed, 12 insertions(+)

diff --git a/src/comm_mpi.f b/src/comm_mpi.f
index 11df316..bf3dc4b 100644
--- a/src/comm_mpi.f
+++ b/src/comm_mpi.f
@@ -107,6 +107,9 @@ c
       real x(n), w(n)
       character*3 op
 c
+#ifdef GEOPM_BARRIER
+      call mpi_barrier(nekcomm, ierr)
+#endif
       if (op.eq.'+  ') then
          call mpi_allreduce (x,w,n,nekreal,mpi_sum ,nekcomm,ierr)
       elseif (op.EQ.'M  ') then
@@ -135,6 +138,9 @@ c
       integer x(n), w(n)
       character*3 op
 
+#ifdef GEOPM_BARRIER
+      call mpi_barrier(nekcomm, ierr)
+#endif
       if     (op.eq.'+  ') then
         call mpi_allreduce (x,w,n,mpi_integer,mpi_sum ,nekcomm,ierr)
       elseif (op.EQ.'M  ') then
@@ -163,6 +169,9 @@ c
       integer*8 x(n), w(n)
       character*3 op
 
+#ifdef GEOPM_BARRIER
+      call mpi_barrier(nekcomm, ierr)
+#endif
       if     (op.eq.'+  ') then
         call mpi_allreduce (x,w,n,mpi_integer8,mpi_sum ,nekcomm,ierr)
       elseif (op.EQ.'M  ') then
diff --git a/test/example1/makenek-intel b/test/example1/makenek-intel
index 1a2733b..affd0e7 100755
--- a/test/example1/makenek-intel
+++ b/test/example1/makenek-intel
@@ -34,6 +34,9 @@ USR_LFLAGS="-qopenmp -mcmodel=medium -shared-intel -L${GEOPM_ROOT}/lib -lgeopm -
 
 # generic compiler flags
 G="-DGEOPM -I${GEOPM_ROOT}/include"
+if [ ! -z "${GEOPM_BARRIER+x}" ]; then
+    G="-DGEOPM_BARRIER ${G}"
+fi
 
 # optimization flags
 OPT_FLAGS_STD="-qopenmp -O3 -g -xCORE-AVX2 -mcmodel=medium -shared-intel"
-- 
2.23.0

